File: core/src/main/java/com/cz/android/sample/function/FunctionManager.java
package com.cz.android.sample.function;

import android.app.Activity;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import com.cz.android.sample.api.SampleItem;
import com.cz.android.sample.extension.ExtensionHandler;
import java.util.ArrayList;
import java.util.List;

/**
 * @author Created by cz
 * @date 2020-01-27 18:08
 * @email bingo110@126.com
 * @see SampleFunction
 */
public class FunctionManager implements ExtensionHandler<SampleFunction> {
    /**
     * all the action plugin
     */
    private List<SampleFunction> functionList = new ArrayList<>();

    @Override
    public boolean handle(@NonNull final String className, @NonNull final String superClass,
            @NonNull final List<String> interfaces) {
        if (interfaces.contains(SampleFunction.class.getName())) {
            try {
                final Class<?> clazz = Class.forName(className);
                SampleFunction function = (SampleFunction) clazz.newInstance();
                register(function);
            } catch (ClassNotFoundException | IllegalAccessException | InstantiationException e) {
                e.printStackTrace();
            }
            return true;
        }
        return false;
    }

    /**
     * Register a plugin. It will put new plugin to list.
     *
     * @param plugin
     */
    @Override
    public void register(SampleFunction plugin) {
        functionList.add(plugin);
    }

    /**
     * unregister a plugin, It will remove the plugin from list.
     *
     * @param plugin
     */
    @Override
    public void unregister(SampleFunction plugin) {
        functionList.remove(plugin);
    }

    public List<SampleFunction> getFunctionList() {
        return functionList;
    }

    /**
     * execute
     *
     * @param context
     * @param item
     */
    public <T extends Activity> void execute(final AppCompatActivity context, final SampleItem item) {
        for (SampleFunction function : functionList) {
            final Class<?> clazz = item.clazz();
            if (function.isAvailable(clazz)) {
                function.execute(context, item);
            }
        }
    }
}

File: core/src/main/java/com/github/jackchen/android/core/function/FunctionManager.kt
package com.github.jackchen.android.core.function

import androidx.appcompat.app.AppCompatActivity
import com.github.jackchen.android.core.extension.ExtensionHandler
import com.github.jackchen.android.sample.api.SampleItem
import java.util.ArrayList

/**
 * @author Created by cz
 * @date 2020-01-27 18:08
 * @email bingo110@126.com
 * @see SampleFunction
 */
object FunctionManager : ExtensionHandler<SampleFunction> {
  private val FUNCTION_CLASS_DESC = SampleFunction::class.java.name.replace('.', '/')

  /**
   * all the action plugin
   */
  private val functionList: MutableList<SampleFunction> = ArrayList()
  override fun handle(
    className: String,
    superClass: String,
    interfaces: List<String>
  ): Boolean {
    if (interfaces.contains(FUNCTION_CLASS_DESC)) {
      try {
        val clazz = Class.forName(className)
        val function = clazz.newInstance() as SampleFunction
        register(function)
      } catch (e: Exception) {
        e.printStackTrace()
      }
      return true
    }
    return false
  }

  /**
   * Register a plugin. It will put new plugin to list.
   *
   * @param plugin
   */
  override fun register(plugin: SampleFunction) {
    functionList.add(plugin)
  }

  /**
   * unregister a plugin, It will remove the plugin from list.
   *
   * @param plugin
   */
  override fun unregister(plugin: SampleFunction) {
    functionList.remove(plugin)
  }

  fun getFunctionList(): List<SampleFunction> {
    return functionList
  }

  /**
   * execute
   *
   * @param context
   * @param item
   */
  fun execute(context: AppCompatActivity, item: SampleItem) {
    for (function in functionList) {
      val clazz = item.clazz()
      if (function.isAvailable(clazz)) {
        function.execute(context, item)
      }
    }
  }
}
